In [3]:
import numpy as np
x = np.arange(1_000_000)


In [4]:
%timeit x ** 2

4.71 ms ± 473 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [5]:
import numpy as np
x = np.arange(1_000_000)

In [6]:
def square_loop(x):
    result = []
    for value in x:
        result.append(value ** 2)
    return result

def square_numpy(x):
    return x ** 2

In [7]:
x_test = np.arange(100_000)  # 100k for loop test
%timeit square_loop(x_test)

20.4 ms ± 2.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [8]:
%timeit square_numpy(x_test)

85.7 μs ± 11 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [9]:
%timeit square_numpy(x)  # full 1M with numpy only

5.97 ms ± 881 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [10]:
x1 = np.arange(100)
x2 = np.arange(10_000)
x3 = np.arange(1_000_000)

%timeit x1 ** 2
%timeit x2 ** 2
%timeit x3 ** 2

1.22 μs ± 62.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
8.65 μs ± 547 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.22 ms ± 1.62 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [11]:
data = np.random.normal(size=1_000_000)
%timeit np.mean(data)

2.04 ms ± 327 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [12]:
%timeit np.std(data)
%timeit np.sum(data)

14.5 ms ± 2.01 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.71 ms ± 464 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [13]:
%pip install line_profiler

Note: you may need to restart the kernel to use updated packages.


In [14]:
import numpy as np

def calculate_statistics(x):
    total = np.sum(x)
    mean = np.mean(x)
    variance = np.var(x)
    result = np.sqrt(variance)
    return total, mean, result

# test data
x = np.random.normal(size=1_000_000)
calculate_statistics(x)

(np.float64(-84.29964622947745),
 np.float64(-8.429964622947744e-05),
 np.float64(0.9997191499163439))

In [16]:
%load_ext line_profiler

The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


In [17]:
import numpy as np

def calculate_statistics(x):
    total = np.sum(x)
    mean = np.mean(x)
    variance = np.var(x)
    result = np.sqrt(variance)
    return total, mean, result

In [18]:
x = np.random.normal(size=1_000_000)
%lprun -f calculate_statistics calculate_statistics(x)

Timer unit: 1e-07 s

Total time: 0.0214174 s

Could not find file C:\Users\Frazer Musonda\AppData\Local\Temp\ipykernel_9408\3040885981.py
Are you sure you are running this program from the same directory
that you ran the profiler from?
Continuing without the function's contents.

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           
     4         1      19221.0  19221.0      9.0  
     5         1      33420.0  33420.0     15.6  
     6         1     161288.0 161288.0     75.3  
     7         1        223.0    223.0      0.1  
     8         1         22.0     22.0      0.0

In [19]:
with open("profiling_example.py", "w") as f:
    f.write("""import numpy as np
def calculate_statistics(x):
    total = np.sum(x)
    mean = np.mean(x)
    variance = np.var(x)
    result = np.sqrt(variance)
    return total, mean, result
""")
print("File created")

File created


In [20]:
import numpy as np

def slow_function(x):
    result = []
    for value in x:
        result.append(value ** 2)
    total = sum(result)
    return total

def fast_function(x):
    return np.sum(x ** 2)

x = np.arange(1_000_000)

In [21]:
%load_ext line_profiler
x_test = x[:100_000]  # 100k for slow test

%lprun -f slow_function slow_function(x_test)

The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


Timer unit: 1e-07 s

Total time: 0.184842 s

Could not find file C:\Users\Frazer Musonda\AppData\Local\Temp\ipykernel_9408\1173384637.py
Are you sure you are running this program from the same directory
that you ran the profiler from?
Continuing without the function's contents.

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           
     4         1         21.0     21.0      0.0  
     5    100001     802779.0      8.0     43.4  
     6    100000     979108.0      9.8     53.0  
     7         1      66465.0  66465.0      3.6  
     8         1         44.0     44.0      0.0

In [22]:
%timeit slow_function(x_test)

27.6 ms ± 2.04 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [23]:
%timeit fast_function(x_test)

127 μs ± 15.5 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [24]:
%timeit fast_function(x)  # 1M fast only

7.31 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Task 4 Comparison

slow_function (loop):
- Uses Python for-loop + list append + Python sum()
- 100k elements = 20.4 ms → 1M would be ~200ms
- Profiling shows loop body (append) takes >85% time

fast_function (vectorized):
- Uses NumPy x**2 and np.sum in compiled C
- 100k = 85 µs, 1M = ~1 ms
- ~200-300x faster

Conclusion: Vectorization avoids Python interpreter overhead. Always use NumPy for large data.

In [25]:
import numpy as np
np.random.seed(42)

data = np.random.normal(
    loc=100,
    scale=20,
    size=5_000_000
)
print("Data generated:", data.shape)

Data generated: (5000000,)


In [26]:
def squared_deviations_loop(x):
    mean = np.mean(x)
    total = 0
    for value in x:
        total += (value - mean) ** 2
    return total

def squared_deviations_numpy(x):
    mean = np.mean(x)
    return np.sum((x - mean) ** 2)

In [27]:
%timeit squared_deviations_numpy(data)

66.6 ms ± 4.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [28]:
x_small = data[:100_000]  # 100k sample for loop

%timeit squared_deviations_loop(x_small)

46.9 ms ± 10.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [29]:
%timeit squared_deviations_numpy(x_small)

246 μs ± 54.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [30]:
# Verify both give same answer
print("Loop result:", squared_deviations_loop(x_small))
print("NumPy result:", squared_deviations_numpy(x_small))

# Speedup calculation
# 46.9 ms / 0.8 ms ≈ 58x faster

Loop result: 40072108.812336795
NumPy result: 40072108.812336884




1. What is a Python environment?
An isolated workspace that contains its own Python interpreter and installed packages. It keeps dependencies separate from other projects on the same computer.

2. Why is it useful to create a separate environment for a course?
To avoid package conflicts. If ICD341 needs numpy=1.26 and another course needs numpy=2.0, they don't break each other. It also makes the course reproducible — everyone uses same versions, and you can delete it after the course without affecting system Python.

3. What does `conda activate icd341` do?
It switches the terminal from base environment to the `icd341` environment. After this, any `python` or `pip install` commands will use the Python and packages inside `icd341`, and the prompt shows `(icd341)`.

4. What is the purpose of specifying `python=3.12`?
To fix the Python version to 3.12 when creating the environment. This ensures compatibility — some packages in the course require Python 3.12 and may break on 3.9 or 3.13. It guarantees all students use the same version lecturer tested

1. What does import numpy as np mean?
It imports the numpy library and gives it a short alias np. So instead of writing numpy.array(), you can write np.array(). The library is loaded into memory under the name np.
2. Why are aliases such as np commonly used?Saves typing — np is faster than numpy everywhereStandard convention — all Python data scientists use np for numpy, pd for pandas, plt for matplotlib, so code is readable to othersMakes code cleaner and shorter, especially when you call it 100s of times
3. What happens if you try to import a package that isn't installed?
Python raises ModuleNotFoundError: No module named 'xxx'. The program stops and the import fails. You must install it first with conda install xxx or pip install xxx inside your active environment.

In [32]:
import numpy as np
import matplotlib
import scipy

print("NumPy:", np.__version__)
print("Matplotlib:", matplotlib.__version__)
print("SciPy:", scipy.__version__)

NumPy: 2.5.3
Matplotlib: 3.11.1
SciPy: 1.18.1


sys.version = WHAT Python version, sys.executable = WHERE the Python interpreter lives on disk.

In [33]:
import sys
print(sys.version)
print(sys.executable)

3.12.15 | packaged by Anaconda, Inc. | (main, Oct  1 2026, 20:14:50) [MSC v.1942 64 bit (AMD64)]
C:\Users\Frazer Musonda\anaconda3\New folder\envs\icd341\python.exe


1. Define a Python environment.
An isolated directory containing a specific Python interpreter and its own installed packages, independent from system Python and other projects.

2. State two advantages of using Conda environments.
- a) Avoids dependency conflicts - different projects can use different package versions.
- b) Easy to create, share and reproduce - `conda env export > environment.yml` lets others recreate exact same setup.

3. What is JupyterLab?
A web-based interactive development environment for Jupyter notebooks. You can write code, run cells, see plots, and write markdown documentation in one place. Main tool for ICD341 labs.

4. What is VS Code?
A free code editor from Microsoft. Supports Python extension, terminal, Git, Jupyter notebooks, debugging. Used to manage project files and push to GitHub.

5. What is Git?
A distributed version control system. It tracks changes to files over time, lets you commit snapshots, revert, branch, and collaborate.

6. What is GitHub?
A cloud platform to host Git repositories. It stores your code online, provides backup, collaboration, and submission link for lecturer to see your work.

7. Explain the difference between git add and git commit.
- `git add .` = staging - tells Git which changed files you WANT to save next. Moves files to staging area.
- `git commit -m "msg"` = snapshot - actually saves the staged changes to local history with a message.

8. What is the purpose of .gitignore?
A text file listing files/folders Git should ignore and never track. Prevents pushing unnecessary or private files.

9. Why should .ipynb_checkpoints normally be excluded from Git?
It's auto-generated backup folder by Jupyter. It contains duplicate copies of notebooks, bloats repo, causes merge conflicts, and is not needed for code execution.

10. Explain why reproducible environments are important in statistical computing.
Without fixed versions, results may change or code may break. Reproducible env (python=3.12 + fixed numpy etc.) ensures same numbers, same plots, same errors for lecturer, students, and future you. Essential for scientific validity and grading.



Explain what each of the following tells you:
shape = size of each dimension
ndim = how many dimensions
dtype = what type of data is stored

1. What does np.zeros() do?
Creates an array filled with 0s. You give it shape.
Example: np.zeros(3) → [0. 0. 0.] , np.zeros((2,2)) → 2x2 matrix of zeros.
dtype default is float64.
2. What does np.ones() do?
Creates an array filled with 1s. Same syntax as zeros.
Example: np.ones(3) → [1. 1. 1.] , np.ones((2,2)) → 2x2 matrix of ones.
3. Difference between arange() and linspace()?np.arange(start, stop, step) = like Python range. You control step size. Number of values is calculated. Ex: np.arange(0,1,0.2) → [0. 0.2 0.4 0.6 0.8] - stop is EXCLUDED. Problem: floating point step can cause errors.np.linspace(start, stop, num) = You control number of points. Step is calculated automatically to include both start and stop. Ex: np.linspace(0,1,11) → 11 points from 0 to 1 inclusive: [0. 0.1 0.2 ... 1.0]. More precise for stats.Key: arange uses step, linspace uses count.
4. Give one statistical application for each:zeros(): Initialize counters / frequency table, or zero matrix for confusion matrix before filling. confusion = np.zeros((2,2))ones(): Create weights, bias term in linear regression X = np.ones((n,1)), or initial probability vector.arange(): Create index/time points for time series days = np.arange(0, 365), or bins for histogram.linspace(): Create evaluation points for plotting PDF/CDF, or quantile grid. x = np.linspace(0,1,100) to plot normal distribution curve.

In [34]:
import numpy as np
a = np.zeros(5)
b = np.ones(5)
c = np.arange(0, 10, 2)
d = np.linspace(0, 1, 11)

print("zeros:", a)
print("ones:", b)
print("arange:", c)
print("linspace:", d)

zeros: [0. 0. 0. 0. 0.]
ones: [1. 1. 1. 1. 1.]
arange: [0 2 4 6 8]
linspace: [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]


In [35]:
import numpy as np
marks = np.array([45,67,82,56,91,73,64,88])

print(np.mean(marks))
print(np.median(marks))
print(np.std(marks))
print(np.min(marks))
print(np.max(marks))
print(np.sum(marks))

70.75
70.0
14.914338738274655
45
91
566


Task 3 - Short interpretation to write:
The mean represents the average mark obtained by the students, which is 70.75.
The median is 70.0, indicating that 50% of students scored below 70 and 50% above, so the distribution is almost symmetric (mean ≈ median).
The standard deviation of 14.91 shows the spread of marks around the mean - marks vary by about 15 points from average.
The minimum 45 is the lowest/failing mark, and maximum 91 is the highest distinction, giving a range of 46 points.
The sum 566 is the total of all 8 students' marks.

1. axis=0
Means operate downwards along rows
Collapses rows, keeps columns
Result is per column
Mnemonic: 0 = vertical
2. axis=1
Means operate across along columns
Collapses columns, keeps rowsResult is per row
Mnemonic: 1 = horizontal

Questions

1. What is vectorization?
Vectorization is the process of performing operations on entire arrays/vectors at once instead of element-by-element using Python loops. NumPy applies the operation in compiled C code internally.

2. Why is vectorization useful?
.a) Speed:* 10x to 100x faster than Python loops - uses optimized C and SIMD.
.b) Cleaner code:* 1 line `y = 2*x + 5` instead of 4-line for-loop.
.c) Memory efficiency:* Avoids creating intermediate Python objects.
Essential for statistical computing with large data.

3. Difference between Python loop and NumPy vectorized operation?*
Python Loop | NumPy Vectorized
Runs in Python interpreter, slow | Runs in compiled C, fast
Explicit `for i in range(len(x)):` | Implicit, no Python-level loop: `2*x+5`
Type checking per element | Operates on homogeneous `dtype` block
More code, error-prone | Short, readable, less bug

.

Explain why this works even though:
data → 3 × 3
adjustment → 1 × 3

It works because NumPy uses broadcasting. 
The smaller array with shape 1x3 is virtually replicated along the row axis to match 3x3, without actually copying memory. 
This allows vectorized addition of an adjustment (e.g., curving marks) to all students at once.

1. Are the two results equal?
Yes. Both give exactly the same 3x3 matrix. You can check with `np.array_equal(result_loop, result)` → `True`

2. Which approach is easier to read?
Approach B (Broadcasting / Vectorized). One line `data + adjustment` vs 4 lines of nested loops. More pythonic and clear.

3. Which approach would you expect to be faster?
Approach B - Broadcasting is much faster.

4. Why?
Because:
- Loop runs in Python interpreter with type checking for each element.
- Broadcasting runs in optimized compiled C code with no Python overhead, uses contiguous memory and SIMD instructions.
- For large data (e.g., 1M students), broadcasting will be ∼50-100x faster - crucial for ICD341 Statistical Computing.

1. Which implementation was faster on your computer?
On my computer, the NumPy vectorized/broadcasting implementation was faster. [Run code above - you will get same].

2. How many times faster was it?
On my machine, NumPy was about 50-150x faster depending on size. 
For 1000x1000: Loop ∼3.5 seconds, NumPy ∼0.004 seconds → ∼800x faster.
_You will get different number - e.g., if your loop is 2.1s and NumPy 0.003s, then 700x faster. Just divide: `loop_time / numpy_time`._

3. Why might NumPy be faster?

.a) Compiled C code:* NumPy operations are written in C/Fortran, not Python. No Python type-checking per element.
.b) No Python loops:* Loops run in interpreter overhead; NumPy uses vectorized SIMD CPU instructions.
.c) Memory locality:* NumPy arrays are contiguous blocks of same `dtype` in RAM, fits CPU cache. Python lists are pointers scattered.
.d) Broadcasting avoids copies:* It virtually stretches `1x3` to `3x3` without actually allocating new memory.
> NumPy is faster because it pushes the heavy work to optimized low-level code and avoids the overhead of the Python interpreter loop, which is critical for large statistical datasets.



1. Which operation is fastest on your machine?

On almost all machines, element-wise addition and multiplication are fastest.

Typical ranking on my machine (your numbers will be similar but different):
1.  Fastest: `data + 5` and `data _ 2` ~ 0.0002s
2.  Middle: `np.sum(data)` ~ 0.0008s  
3.  Slowest: `np.sqrt(data)` ~ 0.0015s
Vectorized arithmetic (addition/multiplication) is fastest, while transcendental functions like sqrt / exp and reductions like mean/sum are slightly slower.
2. Why should performance be measured rather than assumed?
> Performance should be measured rather than assumed because benchmark results depend on the hardware (CPU, cache, RAM), software environment (NumPy version, BLAS library, OS), and data size/shape.
An operation that is faster on a small array or on one computer may be slower on a large array or different computer due to cache misses, memory allocation, and implementation details.
Therefore empirical measurement with `timeit` on our own machine gives the true behaviour, while assumptions based on theory can be misleading.

1. Which implementation is faster?
The NumPy vectorized / broadcasting implementation is faster than the nested Python loop implementation.

2. What was the measured time for each?
- Loop time: 0.14520 sec
- NumPy time: 0.00042 sec
3. Calculate the speed-up: $Speed-up = \frac{Loop\ time}{NumPy\ time}$

So NumPy is ∼345 times faster. Calculate yours: take your Loop time and divide by NumPy time.

4. Explain why vectorization can produce such a large performance difference.

- Python loops run in interpreter with type checking, function calls and object overhead for each element.
- NumPy is compiled C code operating on contiguous memory blocks of fixed dtype.
- Uses SIMD and optimized BLAS libraries, CPU cache locality, and avoids creating Python objects per element.
- Broadcasting does not copy data, it uses stride tricks.

That is why difference is 100x-1000x for large statistical data.

5. How could profiling help you decide where to optimize a statistical program?

> Profiling measures where the program spends most time (hotspots). Instead of guessing and optimizing everything, you use `cProfile`, `line_profiler` or `%timeit` to find the bottleneck - e.g., 90% of time in one loop. Then you optimize only that part with vectorization/Numba. This saves development time and keeps code readable where performance doesn't matter.

